tools/static-code-analysis/src/main/java/org/eclipse/smarthome/tools/analysis/tools/CheckstyleChecker.java

```diff
@@ -0,0 +1,99 @@
+import org.apache.maven.plugins.annotations.ResolutionScope;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+
+/**
+ * Executes the
+ * <a href="https://maven.apache.org/components/plugins/maven-checkstyle-plugin/">maven-checkstyle-plugin</a> with a
+ * predefined ruleset file and configuration properties
+ *
+ * @author Svilen Valkanov
+ *
+ */
+@Mojo(name = "checkstyle", requiresDependencyResolution = ResolutionScope.COMPILE)
+public class CheckstyleChecker extends AbstractChecker {
+
+    /**
+     * The type of the ruleset that will be used
+     */
+    @Parameter(property = "ruleset", defaultValue = "bundle")
+    protected String rulesetType;
+
+    /**
+     * The version of the maven-checkstyle-plugin that will be used
+     */
+    @Parameter(property = "maven.checkstyle.version", defaultValue = "2.17")
+    private String checkstyleMavenVersion;
+
+    /**
+     * Location of the properties file that contains configuration options for the maven-checkstyle-plugin
+     */
+    private static final String CHECKSTYLE_PROPERTIES_FILE = "configuration/checkstyle.properties";
+
+    /**
+     * Directory where the ruleset files for checkstyle are located
+     */
+    private static final String CHECKSTYLE_RULESET_DIR = "rulesets/checkstyle";
+
+    // information about the maven-checkstyle-plugin
+    private static final String MAVEN_CHECKSTYLE_PLUGIN_GOAL = "checkstyle";
+    private static final String MAVEN_CHECKSTYLE_PLUGIN_ARTIFACT_ID = "maven-checkstyle-plugin";
+    private static final String MAVEN_CHECKSTYLE_PLUGIN_GROUP_ID = "org.apache.maven.plugins";
+
+    /**
+     * This is a property in the maven-checkstyle-plugin that is used to describe the path to the ruleset file used from
+     * the plugin. It can not be set in the checkstyle.properties file as it depends on the {@link #rulesetType}
+     */
+    private static final String CHECKSTYLE_RULESET_USER_PROPERTY = "checkstyle.config.location";
+
+    @Override
+    public void execute() throws MojoExecutionException {
+
+        Log log = getLog();
+        ClassLoader cl = getMavenRuntimeClasspathClassLoader();
+        Properties userProps = loadPropertiesFromFile(cl, CHECKSTYLE_PROPERTIES_FILE);
+
+        // Gets the absolute path to the ruleset file and sets the value of the corresponding user property
+        String ruleset = CHECKSTYLE_RULESET_DIR + "/" + rulesetType + ".xml";
+
+        String key = CHECKSTYLE_RULESET_USER_PROPERTY;
+        URL rulesetLocation = cl.getResource(ruleset);
+        String absoluteLocation = rulesetLocation.toString();
+        log.debug("Config file found at " + absoluteLocation);
+        userProps.setProperty(key, absoluteLocation);
+
+        // Add the static-code-analysis plugin as dependency to maven-checkstyle-plugin, because this plugin contains
+        // custom checks
+        Dependency dep = dependency(SMARTHOME_TOOLS_GROUP_ID, SMARTHOME_TOOLS_ARTIFACT_ID, mavenProject.getVersion());
```

Comment:
This is a subtle bug: This will use the maven version of the project which is using this plugin - I e.g. used it on openHAB2-addons, version 2.1.0-SNAPSHOT, which made the plugin fail with an error (since it tried to find static-code-analysis in version 2.1.0-SNAPSHOT instead of 0.9.0-SNAPSHOT).

Same bug for the other mojos.